### PR TITLE
feat(outputs/wavefront): make maximum http batch size configurable

### PR DESCRIPTION
**Pull request by LukeWinikates**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

The Wavefront golang SDK can send metrics via http(s) or tcp. In the case of http, the Wavefront SDK batches metrics internally, flushing them periodically or when they reach a maximum batch size. At present, configuring this batch size is not possible from the wavefront plugin output, so the sdk default value (10,000) is always used. In some cases smaller or larger batch sizes might be preferred, or users might want to match the wavefront client's batch size to the batch size configured for the telegraf output.

This pr exposes the wavefront internal batch size for configuration as `http_batch_size`.

I added a unit test validating that the intended default is set, although the `outputs.Outputs` map doesn't necessarily seem like an interface that wants to be used for this purpose.

### CODE REVIEW

**Comment by Hipska** [comment]
I'm wondering if it would make sense to match always with the telegraf configured batch size, so there's no additional config option to be set.

It would also remove the confusion of what will happen when `metric_batch_size` > `http_batch_size` or `metric_batch_size` < `http_batch_size`.

**Comment by LukeWinikates** [comment]
> I'm wondering if it would make sense to match always with the telegraf configured batch size, so there's no additional config option to be set.

I think that's reasonable. I took a look at doing this, but I didn't see a way to get at the `metric_batch_size` configuration from within the output plugin itself. Without refactoring the way output plugins are initialized, is there a way to find the `metric_batch_size` from the `Connect` method?

Another thing that could work would be to set a larger default batch size for the wavefront SDK while leaving it non-configurable. Right now the Wavefront output's `immediate_flush` boolean option will cause a batch to be sent on every `Write` call, and I assume `Write` gets called once per Telegraf output batch. If metric_batch_size > 10,000, currently we would cap the batch size at 10,000, but if we change the Wavefront max batch size to to 50,000, then we could honor larger values of `metric_batch_size`. But it would make the batch size bigger for anyone not using `immediate_flush = true`.

I am conflicted about it because I like the transparency of exposing the same configuration hooks that the underlying SDK has, but I would also like the configuration to be simpler.

**Comment by LukeWinikates** [comment]
> Without refactoring the way output plugins are initialized, is there a way to find the `metric_batch_size` from the `Connect` method?

@Hipska thanks for your earlier comment on this PR. Do you happen to know the answer to the above question?

**Comment by Hipska** [comment]
Good question, I checked the code and found a few things:

1. It's currently not possible to know this in the `Connect` method. (Didn't try this, but you could try to add a struct field with the same TOML name. I don't know if this would conflict or not. However, You would still not be able to retrieve the global agent batch size.)
2. The used `wavefront.NewDirectSender` and `wavefront.NewProxySender` methods are deprecated according to their documentation. Advised is to use the `wavefront.NewSender(url)` method.
3. The batch size can be derived from the amount of metrics you get in the `Write` method, it will be less than or equal to the batch size.
4. There seems to be already a possibility to match the metrics batch size with the http batch size for wavefront sdk, by setting the config item of `immediate_flush` to true, which also seems to be the default value, so I don't even know what problem you are actually trying to fix here???

**Comment by LukeWinikates** [comment]
Just curious, is there anything I can follow up on to help this get merged?

I recognize that the proposed change is solving for an uncommon edge case, but it would allow my team to more easily experiment with different batch sizes.

That said, if y'all think it's better not to merge this change, I do think we can eliminate this use case by redesigning the configuration schema of the wavefront plugin and how it initializes the WF client. (I just didn't think that a schema redesign or big internal  changes to the plugin were a great place to start with a first PR).

I also feel bad for @-ing @Hipska over and over again but I'm not sure who else might be reviewing and commenting on PRs.

**Review by Hipska — APPROVED** [review]

File: plugins/outputs/wavefront/README.md [hunk 1]
```diff
@@ -15,6 +15,9 @@ Wavefront data format over TCP.
   ## Authentication Token for Wavefront. Only required if using Direct Ingestion
   #token = "DUMMY_TOKEN"
 
+  ## Number of metrics to send per batch for Direct Ingestion. Ignored unless 'url' is set. Default is 10,000. Values higher than 40,000 are not recommended.
```
[the remark this answers is not in the document]
Comment (powersj) [reply plugins/outputs/wavefront/README.md]: Interesting it should have, assuming you have the latest commits from master.

Thanks for the URL! I'm going to land this.

**Pull request merged**